contrib/babelfishpg_tsql/runtime/functions.c

```diff
@@ -1564,3 +1565,72 @@ smallint_radians(PG_FUNCTION_ARGS)
 
 	PG_RETURN_INT32((int32)result);
 }
+
+/* Returns the database schema name for schema-scoped objects. */
+Datum 
+object_schema_name(PG_FUNCTION_ARGS)
+{	
+	Oid object_id;
+	Oid database_id;
+	Oid	user_id = GetUserId();
+	Oid namespace_oid = InvalidOid;
+	Oid relid = InvalidOid;
+	char* namespace_name;
+	const char* schema_name;
+
+	if(PG_ARGISNULL(0))
+		PG_RETURN_NULL();
+	else
+		object_id = (Oid) PG_GETARG_INT32(0);
+
+	if(PG_ARGISNULL(1))
+		database_id = get_cur_db_id();
+	else {
+		database_id = (Oid) PG_GETARG_INT32(1);
+		user_id = GetSessionUserId();
+	}
+
+	/* if user inputs invalid database_id */
+	if (!DbidIsValid(database_id))
+		PG_RETURN_NULL();
+	
+	/* lookup namespace_oid in pg_class */
+	relid = get_rel_namespace(object_id);
+	if(OidIsValid(relid) && 
+		pg_class_aclcheck(object_id, user_id, ACL_SELECT) == ACLCHECK_OK)
```

Comment:
why are we only doing aclcheck for relid and also continuing if check fails?

Reply:
We need to check if the user has permission to view the object, otherwise we output NULL.
But there is no need to continue if the check fails since we have found the object already. Updated it in the latest commit.